File: fpscribe.py
from collections import namedtuple
import logging
import struct

log = logging.getLogger(__name__)


def main(argv, usb_hids):
    hid_dev = argv[1]
    fp = FootPedal(usb_hids.sub_rd(hid_dev))
    for e in fp.each_event():
        log.debug('event: %s', e)


class FootPedal(object):
    '''VEC USB Footpedal (USB device idVendor=05f3, idProduct=00ff)
    registers as hid device.

    ack: footpedal__ explains that each pedal action generates three
    groups of 8 characters.

    __ https://code.google.com/p/footpedal/
    '''

    pedal_qty = 3
    pedal_data_fmt = 'B3xB3x'
    pedal_data_size = struct.calcsize(pedal_data_fmt)
    event_size = pedal_data_size * pedal_qty

    def __init__(self, device):
        def each_event():
            stream = device.open_rd()
            while 1:
                data = stream.read(self.event_size)
                if not data:
                    break
                yield self.unpack(data)

        self.each_event = each_event

    @classmethod
    def unpack(cls, data):
        r'''Unpack footpedal event data.

        >>> chunks = [chr(ix) + '\x00\x00\x00\x01\x00\x00\x00'
        ...           for ix in range(1, 4)]
        >>> data = ''.join(chunks)
        >>> FootPedal.unpack(data)
        ... # doctest: +NORMALIZE_WHITESPACE
        [PedalEvent(ix=1, pressed=1),
         PedalEvent(ix=2, pressed=1),
         PedalEvent(ix=3, pressed=1)]
        '''
        pedals_data = [
            struct.unpack(cls.pedal_data_fmt,
                          data[ix:ix + cls.pedal_data_size])
            for ix in range(0, cls.event_size, cls.pedal_data_size)]
        return [PedalEvent(ix, status)
                for (ix, status) in pedals_data]


PedalEvent = namedtuple('PedalEvent', ['ix', 'pressed'])


# Read access rights.
Rd = namedtuple('Rd', ['path', 'sub_rd', 'open_rd'])


def hid_devs(openf,
             path='/dev/usb/hiddev'):
    def sub_rd(n):
        if not n.isdigit():
            raise IOError('not a device number: ' + n)
        return Rd(path + n, sub_rd=_cannot,
                  open_rd=lambda: openf(path + n))

    return Rd(path, sub_rd, open_rd=_cannot)


def _cannot(*args):
    raise IOError('not authorized: %s' % args)


def argv_rd(argv, rd):
    '''Attenuate a rd so only CLI args are allowed as sub paths.
    '''
    def sub_rd(n):
        if n not in argv:
            raise IOError('not CLI arg: ' + n)
        return rd.sub_rd(n)
    return Rd(rd.path, sub_rd, rd.open_rd)


if __name__ == '__main__':
    def _logging(level=logging.DEBUG):
        logging.basicConfig(level=level)

    def _with_caps():
        from __builtin__ import open as openf
        from sys import argv

        hid_rd = hid_devs(openf)
        main(argv, argv_rd(argv, hid_rd))

    _logging()
    _with_caps()

File: fpscribe.rs
#![feature(phase)]
#[phase(plugin, link)] extern crate log;

use std::io;
use std::io::{IoResult, IoError, Reader};


fn cap_main(args: || -> Vec<String>, usb_hids: Box<Rd>) {
    let dev_num = args()[1].clone();
    match usb_hids.sub_rd(dev_num.as_slice()) {
        Err(why) => fail!(format!("{}", why)),
        Ok(hid) => {
            let fp: FootPedal = FootPedal::new(hid);
            loop {
                let e = fp.events.recv();
                debug!("event: {}", e)
            }
        }
    }
}


struct FootPedal {
    events: Receiver<Vec<PedalEvent>>
}


static pedal_ixs: [int, ..3] = [1, 2, 3];
static pedal_data_size: uint = 8;
static event_size: uint = pedal_data_size * 3;

impl FootPedal {
    //! VEC USB Footpedal (USB device idVendor=05f3, idProduct=00ff)
    //! registers as hid device.
    //!
    //! ack: [footpedal]() explains that each pedal action generates three
    //! groups of 8 characters.
    //!
    //! [footpedal]: https://code.google.com/p/footpedal/

    fn new(device: Box<Rd+Send>) -> FootPedal {
        let (tx, rx) = channel();

        spawn(proc() {
            let mut stream = device.open_rd().unwrap();
            loop {
                match stream.read_exact(event_size) {
                    Ok(data) => tx.send(FootPedal::unpack(data)),
                    Err(why) => fail!("{}", why)
                }
            }
        });

        FootPedal{ events: rx }
    }

    /** Unpack footpedal event data.

    >>> chunks = [chr(ix) + '\x00\x00\x00\x01\x00\x00\x00'
    ...           for ix in range(1, 4)]
    >>> data = ''.join(chunks)
    >>> FootPedal.unpack(data)
    ... # doctest: +NORMALIZE_WHITESPACE
    [PedalEvent(ix=1, pressed=1),
     PedalEvent(ix=2, pressed=1),
     PedalEvent(ix=3, pressed=1)]
    */
    fn unpack(data: Vec<u8>) -> Vec<PedalEvent> {
        let each = |n: &int| {
            let ix = (n - 1) as uint;
            PedalEvent{ ix: data[ix * pedal_data_size],
                        pressed: data[ix * pedal_data_size + 4] }
        };
        pedal_ixs.iter().map(each).collect()
    }
}


#[deriving(Show)]
struct PedalEvent {
    ix: u8,
    pressed: u8  // TODO: bool
}


// Read access rights.
trait Rd {
    fn full_path(&self) -> String;
    fn sub_rd(&self, n: &str) -> IoResult<Box<Rd+Send>>;
    fn open_rd(&self) -> IoResult<Box<Reader+Send>>;
}


struct HidDevs {
    dev_usb: Box<Rd+Send>
}

impl HidDevs {
    fn new(fs: Box<Rd>) -> IoResult<HidDevs> {
        fs.sub_rd("/dev/usb").map( |rd| HidDevs { dev_usb: rd } )
    }
}

impl Rd for HidDevs {
    fn sub_rd(&self, n: &str) -> IoResult<Box<Rd+Send>> {
        if n.char_len() == 1 && n.char_at(0).is_digit() {
            self.dev_usb.sub_rd(("hiddev".to_string() + n.to_string()).as_slice())
        } else {
            Err(IoError{ desc: "not a device number",
                         detail: Some(n.to_string()),
                         kind: io::PermissionDenied })
        }
    }

    fn full_path(&self) -> String {
        self.dev_usb.full_path()
    }

    fn open_rd(&self) -> IoResult<Box<Reader+Send>> {
        Err(IoError{ desc: "cannot open directory",
                     detail: None,
                     kind: io::MismatchedFileTypeForOperation })
    }
}


struct ArgvRd {
    argv: Vec<String>,
    rd: Box<Rd+Send>
}

impl Rd for ArgvRd {
    fn sub_rd(&self, n: &str) -> IoResult<Box<Rd+Send>> {
        if self.argv.iter().any(|s| s.as_slice() == n) {
            self.rd.sub_rd(n)
                } else {
            Err(IoError{ desc: "not CLI arg",
                         detail: Some(n.to_string()),
                         kind: io::PermissionDenied })
        }
    }

    fn full_path(&self) -> String {
        self.rd.full_path()
    }

    fn open_rd(&self) -> IoResult<Box<Reader+Send>> {
        self.rd.open_rd()
    }
}


mod trusted {
    use std::io::IoResult;
    use std::io::fs::File;

    use super::Rd;

    pub struct PosixFile {
        pub path: String
    }

    impl ::Rd for PosixFile {
        fn full_path(&self) -> String {
            self.path.clone()
        }

        fn sub_rd(&self, n: &str) -> IoResult<Box<Rd+Send>> {
            let sub_path = self.path + "/" + n; // TODO: real path join
            Ok(box PosixFile{ path: sub_path} as Box<Rd+Send>)
        }

        fn open_rd(&self) -> IoResult<Box<Reader+Send>> {
            let rf: IoResult<File> = File::open(&Path::new(self.path.clone()));
            rf.map(|f| box f as Box<Reader+Send>)
        }
    }
}

fn main() {
    use std::os::args;

    let root_rd: Box<Rd> = box ::trusted::PosixFile{ path: "/".to_string() };
    let hid_devs = box HidDevs::new(root_rd).unwrap() as Box<Rd+Send>;
    let arg_dev_rd = box ArgvRd{ argv: args(), rd: hid_devs } as Box<Rd>;

    cap_main(args, arg_dev_rd)
}
